pinot-tools/src/main/java/org/apache/pinot/tools/PinotDataAndQueryAnonymizer.java

```diff
@@ -530,68 +544,107 @@ private void generateDerivedValuesForGlobalDictionary(OrigAndDerivedValueHolder
         break;
       default:
         throw new UnsupportedOperationException("global dictionary currently does not support: " + valueHolder._dataType.name());
     }
   }
 
-  private Integer[] generateDerivedIntValues(int cardinality) {
+  private Integer[] generateDerivedIntValuesForGD(int cardinality) {
     Integer[] values = new Integer[cardinality];
     for (int i = 0; i < cardinality; i++) {
       values[i] = INT_BASE_VALUE + i;
     }
     return values;
   }
 
-  private Long[] generateDerivedLongValues(int cardinality) {
+  private Long[] generateDerivedLongValuesForGD(int cardinality) {
     Long[] values = new Long[cardinality];
     for (int i = 0; i < cardinality; i++) {
       values[i] = LONG_BASE_VALUE + (long)i;
     }
     return values;
   }
 
-  private Float[] generateDerivedFloatValues(int cardinality) {
+  private Float[] generateDerivedFloatValuesForGD(int cardinality) {
     Float[] values = new Float[cardinality];
     for (int i = 0; i < cardinality; i++) {
       values[i] = FLOAT_BASE_VALUE + (float)i;
     }
     return values;
   }
 
-  private Double[] generateDerivedDoubleValues(int cardinality) {
+  private Double[] generateDerivedDoubleValuesForGD(int cardinality) {
     Double[] values = new Double[cardinality];
     for (int i = 0; i < cardinality; i++) {
       values[i] = DOUBLE_BASE_VALUE + (double)i;
     }
     return values;
   }
 
-  private String[] generateDerivedStringValues(OrigAndDerivedValueHolder valueHolder) {
+  private String[] generateDerivedStringValuesForGD(OrigAndDerivedValueHolder valueHolder) {
     int cardinality = valueHolder._index;
     String[] values = new String[cardinality];
     for (int i = 0; i < cardinality; i++) {
-      int origValLength = ((String)valueHolder._origValues[i]).length();
+      String val = (String)valueHolder._origValues[i];
+      int origValLength;
+      if (val == null || val.equals("") || val.equals(" ") || val.equals("null")) {
+        // can't map to null since sort will rearrange and two undesirable cases are possible:
+        // (1) a non-null original value can map to a null derived value in the GD
+        // (2) a null original value can map to a non-null derived value in the GD
+        // so it is better to map both null and non-null to non-null value
+        // arbitrary size chosen as 10
+        origValLength = 10;
+      } else {
+        origValLength = val.length();
+      }
       values[i] = RandomStringUtils.randomAlphanumeric(origValLength);
     }
     Arrays.sort(values);
     return values;
   }
 
+  private ByteArray[] generateDerivedByteValuesForGD(OrigAndDerivedValueHolder valueHolder) {
+    int cardinality = valueHolder._index;
+    ByteArray[] values = new ByteArray[cardinality];
+    Random random = new Random();
+    for (int i = 0; i < cardinality; i++) {
+      ByteArray byteArray = (ByteArray)valueHolder._origValues[i];
+      int origValLength;
+      if (byteArray == null || byteArray.length() == 0) {
```

Comment:
same as before for Strings, leave it at empty byte array

Reply:
done